circbuf.go

```diff
@@ -65,21 +73,34 @@ func (b *Buffer) TotalWritten() int64 {
 }
 
 // Bytes provides a slice of the bytes written. This
-// slice should not be written to.
+// slice should not be written to. The underlying array
+// may point to data that will be overwritten by a subsequent
+// call to Bytes. It does no allocation.
 func (b *Buffer) Bytes() []byte {
 	switch {
 	case b.written >= b.size && b.writeCursor == 0:
 		return b.data
 	case b.written > b.size:
-		out := make([]byte, b.size)
-		copy(out, b.data[b.writeCursor:])
-		copy(out[b.size-b.writeCursor:], b.data[:b.writeCursor])
-		return out
+		copy(b.out, b.data[b.writeCursor:])
+		copy(b.out[b.size-b.writeCursor:], b.data[:b.writeCursor])
+		return b.out
 	default:
 		return b.data[:b.writeCursor]
 	}
 }
 
+// Gets a single byte out of the buffer, at the given position.
+func (b *Buffer) Get(i int64) (byte, error) {
+	switch {
+	case i >= b.written || i >= b.size:
+		return 0, fmt.Errorf("Index out of bounds: %v", i)
+	case b.written > b.size:
```

Comment:
It's not word having the extra case to avoid an addition and a module. Conditionals are more expensive anyway. I'd rewrite to:

```golang
func (b *Buffer) Get(i int64) (byte, error) {
    if i >= b.written || i >= b.size {
        return 0, fmt.Errorf("Index out of bounds: %v", i)
    }
    return b.data[(b.writeCursor + i) % b.size], nil
}
```

Reply:
@petrosagg this is not correct, the code you pasted will access an offset that is outside the bounds of the readable buffer when the buffer has not rolled over -- its bounds are `[0, b.writeCursor)`, so its `i`th byte is accessed with `b.data[i]` as a normal array. When it rolls over and starts overwriting previous data, `b.writeCursor` essentially becomes the array's zero index, so indeed accessing the `i`th byte is `b.writeCursor + i`, but we do have to treat the two cases separately.

Reply:
Oops. Yeah I got it wrong. Nevermind